pkg/apis/core/v1alpha1/portforward_types.go

```diff
@@ -76,6 +79,50 @@ type PortForwardSpec struct {
 var _ resource.Object = &PortForward{}
 var _ resourcestrategy.Validater = &PortForward{}
 
+func NewPortForward(localPort, containerPort int, host string, podID string, ns string, mName string) *PortForward {
+	// generate a consistent name for a port forward based on its component parts
+	name := pfName(localPort, containerPort, host, podID, ns, mName)
+
+	return &PortForward{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Annotations: map[string]string{
+				// Name of the manifest that this Port Forward corresponds to
+				// (we need this to route the logs correctly)
+				AnnotationManifest: mName,
+			},
+		},
+		Spec: PortForwardSpec{
+			PodName:       podID,
+			Namespace:     ns,
+			ContainerPort: containerPort,
+			LocalPort:     localPort,
+			Host:          host,
+		},
+	}
+}
+
+// pfName generates a consistent and unique name for a port forward based on its component parts
```

Comment:
hmmm...i'm not sure this restriction makes much sense, for a couple different reasons:

- we're moving towards a world where people can create their own port forwards. You _could_ try to enforce name and immutability restrictions, but it's not the "easy path", and I don't see a good reason to try to enforce them.

- the real restriction on port forwards is that you can't have two port-forwards binding to the same local port

have you seen how FileWatch handles this? I think a good approach here would be to make the port forward name = manifest name + index (e.g., storybook-1, storybook-2), and do a spec comparison like here: https://github.com/tilt-dev/tilt/blob/master/internal/controllers/core/filewatch/controller.go#L81

- -

Reply:
just basing the index on how many other PFs for that manifest exist? That makes sense but seems like it'd be tricky because I need to know the state AND have PF's stored by manifest in order to make that go, hm.

i might call them "$POD-1" instead because that's at least information I have access to in the `PortForwadSubscriber`--I assume that still fits with the basic suggestion you're making here?

Reply:
(fwiw this naming constraint wouldn't need to exist for user-created PF's / PFs that went straight into the reconciler -- it's specifically for PFs created in the subscriber, so that if we make one for e.g. podA:8080->8080 and the next time around make one for podA:9090-9090, there should be an easy way to check whether the port forward exists or not. this wasn't a problem previously because there was only one entry per pod (which could contain multiple forwards) so it was easy to check if a given pod already had something on it. maybe i'll revert to doing it that way)